core/src/services/chainsafe/core.rs

```diff
@@ -0,0 +1,373 @@
+// specific language governing permissions and limitations
+// under the License.
+
+use std::fmt::Debug;
+use std::fmt::Formatter;
+use std::sync::Arc;
+
+use bytes::Bytes;
+use http::header;
+use http::Request;
+use http::Response;
+use http::StatusCode;
+use serde::Deserialize;
+use serde_json::json;
+use tokio::sync::RwLock;
+
+use crate::raw::*;
+use crate::*;
+
+use super::error::parse_error;
+
+/// Core of [chainsafe](https://storage.chainsafe.io/) services support.
+#[derive(Clone)]
+pub struct ChainsafeCore {
+    pub signer: Arc<RwLock<ChainsafeSigner>>,
+
+    /// The root of this core.
+    pub root: String,
+    /// The api_secret of this core.
+    pub api_secret: String,
+    /// The bucket name of this backend.
+    pub bucket: String,
+
+    pub client: HttpClient,
+}
+
+impl Debug for ChainsafeCore {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Backend")
+            .field("root", &self.root)
+            .field("bucket", &self.bucket)
+            .finish_non_exhaustive()
+    }
+}
+
+impl ChainsafeCore {
+    #[inline]
+    pub async fn send(&self, req: Request<AsyncBody>) -> Result<Response<IncomingAsyncBody>> {
+        self.client.send(req).await
+    }
+
+    pub async fn get_auth_info(&self) -> Result<AuthInfo> {
+        {
+            let signer = self.signer.read().await;
+
+            if !signer.auth_info.bucket_id.is_empty() {
+                let auth_info = signer.auth_info.clone();
+                return Ok(auth_info);
+            }
+        }
+
+        {
+            let mut signer = self.signer.write().await;
+
+            if !signer.auth_info.bucket_id.is_empty() {
+                let auth_info = signer.auth_info.clone();
+                return Ok(auth_info);
+            }
+
+            let req = Request::get("https://api.chainsafe.io/api/v1/buckets")
+                .header(
+                    header::AUTHORIZATION,
+                    format_authorization_by_bearer(&self.api_secret)?,
+                )
+                .body(AsyncBody::Empty)
+                .map_err(new_request_build_error)?;
+
+            let resp = self.send(req).await?;
+
+            let status = resp.status();
+
+            match status {
+                StatusCode::OK => {
+                    let resp_body = &resp.into_body().bytes().await?;
+
+                    let buckets: Vec<Bucket> =
+                        serde_json::from_slice(resp_body).map_err(new_json_deserialize_error)?;
+
+                    for bucket in buckets {
+                        if let Some(name) = bucket.name {
+                            if self.bucket == name {
+                                signer.auth_info.bucket_id = bucket.id;
+                                break;
+                            }
+                        }
+                    }
+
+                    // bucket not found
+                    if signer.auth_info.bucket_id.is_empty() {
```

Comment:
To make everything simpler, how about enforcing users to input bucket_id directly?

Reply:
I don't have any objections, but is it acceptable to call the 'list buckets' API for the bucket_id?

Reply:
> I don't have any objections, but is it acceptable to call the 'list buckets' API for the bucket_id?

No, I don't like this idea; it increases our maintenance burden. I'm ok to add it back in the future if there are real users complain about this.

Reply:
So are we switching to directly inputting the bucket_id?

Reply:
> So are we switching to directly inputting the bucket_id?

Yes, let's do this.

Reply:
fixed